benchmarks/distributed/ddp/compare/python_ddp.py

```diff
@@ -0,0 +1,228 @@
+            return len(self.param_to_offset) == self.ready_param_grad_count
+
+        def empty(self):
+            self.ready_param_grad_count = 0
+
+        def try_hold_param(self, param):
+            """
+            Checks whether current bucket has enough buffer to hold the incoming
+            param. If there is enough space, distribute param into current
+            bucket and Returns true. Otherwise, returns False.
+            """
+            if self.total_elements + param.numel() <= self._MAX_BUFFER_SIZE :
+                self.param_to_offset[param] = self.total_elements
+                self.total_elements += param.numel()
+                return True
+            else:
+                return False
+
+    def __init__(self, module, process_group, async_reduction=True, buffer_size=2 ** 22):
+        super(PythonDDP, self).__init__()
+
+        self.module = module
+        self.process_group = process_group
+        self.world_size = dist.get_world_size(group=self.process_group)
+        self.async_reduction = async_reduction
+
+        # Holds all_reduce handles, used when async_reduction is True
+        self.async_handles = set()
+
+        # Ensure buffer_size is large enough to hold largest param.
+        max_numel = max(p.numel() for p in module.parameters())
+        assert buffer_size > max_numel, "buffer_size: {} should be larger than largest param: {}".format(buffer_size, max_numel)
+
+        # Build buckets for params
+        self.param_to_bucket, self.buckets = self._build_buckets_for_params(buffer_size)
+
+        # Register per-parameter hook to be invoked when grad is ready.
+        for p in self.module.parameters():
+            assert p.requires_grad
+            p.register_hook(functools.partial(self._on_param_grad_ready, p))
+
+    def _build_buckets_for_params(self, max_buffer_size):
+        """
+        Distributes params into list of buckets. Maintains param -> bucket
+        mapping. Returns tuple of (param_to_buckets, buckets).
+        """
+        print("_build_buckets_for_params called")
+        params_to_buckets = {}
+        buckets = set()
+        cur_bucket = self.Bucket(max_buffer_size)
+        total_param = 0
+        for param in self.module.parameters():
+            total_param += 1
+            assert param.requires_grad, "param.requires_grad must be True"
+            if cur_bucket.try_hold_param(param):
+                params_to_buckets[param] = cur_bucket
+                buckets.add(cur_bucket)
+            else:
+                new_bucket = self.Bucket(max_buffer_size)
+                assert new_bucket.try_hold_param(param), "param must be holded in a empty bucket"
+                params_to_buckets[param] = new_bucket
+                buckets.add(new_bucket)
+                cur_bucket = new_bucket
+
+        first_param = next(self.module.parameters())
+        for bucket in buckets:
+            bucket.buffer = first_param.new(bucket.total_elements)
```

Comment:
`torch.Tensor.new` is deprecated, and it is likely to copy the data.

It will more efficient to use `new_empty`. See: https://pytorch.org/docs/stable/generated/torch.Tensor.new_empty.html

Reply:
Good idea, will address it in follow up CLs.